shared-bindings/bleio/UUIDType.c

```diff
@@ -0,0 +1,75 @@
+/*
+ * This file is part of the Micro Python project, http://micropython.org/
+ *
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2018 Artur Pacholec
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include "shared-bindings/bleio/UUIDType.h"
+
+//| .. currentmodule:: bleio
+//|
+//| :class:`UUIDType` -- defines the type of a BLE UUID
+//| =============================================================
+//|
+//| .. class:: bleio.UUIDType
+//|
+//|     Enum-like class to define the type of a BLE UUID.
+//|
+//|     .. data:: TYPE_16BIT
+//|
+//|       The UUID is 16-bit
+//|
+//|     .. data:: TYPE_128BIT
+//|
+//|       The UUID is 128-bit
```

Comment:
It seems weird to have a separate notion of type. Could we actually have two separate classes?

Reply:
One class for the standard, short UUIDs and one for the long custom ones. That way we can check type with standard `isinstance`.

Reply:
Oh like that, I'm not sure if that's the best idea, don't think I ever saw a library do that. If we did it that way then we would need a base class so the other two could inherit from it and we would need to know what kind of uuid we have before creating a class, like this:

```
if uuid_str.len == 36:
    uuid = UUID128(uuid_str)
else:
    uuid = UUID16(uuid_str)
```

If you check other frameworks, they just have one class with usually a type property:
https://infocenter.nordicsemi.com/index.jsp?topic=%2Fcom.nordic.infocenter.s140.api.v6.0.0%2Fstructble__uuid__t.html&resultof="ble_uuid_t"%20"Struct"%20"struct"%20"Reference"%20"refer"%20
http://doc.qt.io/qt-5/qbluetoothuuid.html#minimumSize
https://ianharvey.github.io/bluepy-doc/uuid.html
Even Android uses full blown 128bit UUIDs for both https://developer.android.com/reference/java/util/UUID

Reply:
Most of those are C APIs which can't do duck typing. We don't need to do it just because `bluepy` does.

Ya, off the wire we'll need to do a length check, but the differences between the two will be clearer with two proper types.

* One is standard, 16bit and has a text description. `AssignedName(0x1800, "Generic Access")` Furthermore, instead of providing constants for the ints we can have modules for related names dynamically loaded.
* The other is custom, no description and may be derived from a base ID.

```python
CustomUUID(b'\x4A\x98\x12\x34\x1C\xC4\xE7\xC1\xC7\x57\xF1\x26\x7D\xD0\x21\xE8')
# or support sharing a base id
BASE_UUID = b'\x4A\x98\x00\x00\x1C\xC4\xE7\xC1\xC7\x57\xF1\x26\x7D\xD0\x21\xE8'
CustomUUID(BASE_UUID, 0x1800, first_bit=16)
```

This approach doesn't require a UUIDType enum class or type property. A super class can be used to share code but isn't required.

Reply:
@tannewt Could you explain more in detail what would be the most (circuit)pythonic way to do this and I can fix, I have few more PRs in the oven ;)